### PR TITLE
Move things over from Denali

### PR DESCRIPTION
**Pull request by acburdine**

### CODE REVIEW

**Comment by acburdine** [comment]
This is all done and working (afaict) - just need to add the tests now :)

**Comment by acburdine** [comment]
Done!

File: lib/bootstrap.ts [hunk 1]
```diff
@@ -37,6 +35,7 @@ export default function run(projectPkg?: any)  {
   debug('discovering commands from addons');
   let commands: { [key: string]: typeof Command } = {};
   let coreCommands: { [key: string]: typeof Command };
+  let cliCommands: { [key: string]: typeof Command };
```
Comment (davewasmer) [inline lib/bootstrap.ts]: Can we call this `globalCommands` perhaps? CLI commands feels a tad unspecific - all the commands are CLI commands after all.

File: lib/bootstrap.ts [hunk 2]
```diff
@@ -62,14 +61,11 @@ export default function run(projectPkg?: any)  {
     }
   });
 
-  // Ensure that denali itself is installed so we have the base commands
-  if (!coreCommands) {
-    ui.error('Whoops, looks like you have not installed denali itself yet.');
-    ui.error('You need to install denali globally (`$ npm i -g denali`) alongside the CLI.');
-  }
+  // Special-case denali-cli commands
+  cliCommands = discoverCommands(commands, 'denali-cli', path.join(__dirname, '..', 'commands'));
```
Comment (davewasmer) [inline lib/bootstrap.ts]: Well, that was far less painful than I feared :wink:

File: package.json [hunk 3]
```diff
@@ -25,14 +25,16 @@
...
+    "nyc": "^11.0.2",
     "tslint": "^5.4.3",
     "typescript": "2.2.0"
   },
   "scripts": {
     "build": "tsc && mv dist/bin/denali.js dist/bin/denali && chmod +x dist/bin/denali && echo 'Build successful'",
     "prepublish": "in-publish && npm run build || not-in-publish",
-    "lint": "tslint --project tsconfig.json -c tslint.json --type-check | sed \"s|$PWD/||\"",
-    "test": "npm run lint"
+    "lint": "! tslint --project tsconfig.json -c tslint.json --type-check | sed \"s|$PWD/||\" | grep ERROR",
```
Comment (davewasmer) [inline package.json]: What does this leading `!` do? I'm not familiar with that bash syntax.

File: tsconfig.json [hunk 4]
```diff
@@ -12,9 +12,19 @@
       "inlineSources": true,
       "outDir": "dist",
       "baseUrl": ".",
-      "skipLibCheck": true
+      "skipLibCheck": true,
+      "sourceRoot": ".",
+      "noUnusedLocals": true,
+      "paths": {
+        "denali-cli": [ "lib/index.ts" ]
```
Comment (davewasmer) [inline tsconfig.json]: Why is this needed?

File: package.json (hunk 3 above)
Comment (acburdine) [inline package.json]: It negates the result of the line...e.g. if grep finds nothing, the script exits with 0 because the content is empty. If grep does find something then it exits with 1, halting execution.

File: tsconfig.json (hunk 4 above)
Comment (acburdine) [inline tsconfig.json]: It's probably not...I saw the same thing in Denali and just copied it over, I'm not exactly sure what it does. I'll try removing it and see if I run into any issues.

**Comment by acburdine** [comment]
Should be good now 👍
